service/src/main/java/org/cbioportal/service/impl/CoExpressionServiceImpl.java

```diff
@@ -42,65 +55,171 @@
+        Map<String,List<String>> values = new HashMap<String,List<String>>();
+        for (String entityId : molecularDataMap.keySet()) {
+            List<String> internalValues = new ArrayList<>(Arrays.asList(molecularDataMap.get(entityId).getSplitValues()));
+            values.put(entityId, internalValues);
+        }
+        coExpressionList = computeCoExpressions(values, queryValues, isMolecularProfileBOfGenesetType, threshold);  
+        return coExpressionList;
+    }
+
+    @Override
+    public List<CoExpression> fetchCoExpressions(String geneticEntityId,
+            CoExpression.GeneticEntityType geneticEntityType, List<String> sampleIds, String molecularProfileIdB,
+            String molecularProfileIdA, Double threshold) throws MolecularProfileNotFoundException, GenesetNotFoundException, GeneNotFoundException {
+
+        if (molecularProfileIdA.equals(molecularProfileIdB)) {
+            return fetchCoExpressions(molecularProfileIdA, sampleIds, geneticEntityId, geneticEntityType, threshold);
+        }
 
         List<CoExpression> computedCoExpressions = null;
-        List<? extends MolecularData>  molecularDataListA = null;
-        List<? extends MolecularData>  molecularDataListB = null;
+        List<? extends MolecularData> molecularDataListA = null;
+        List<? extends MolecularData> molecularDataListB = null;
         if (geneticEntityType.equals(GeneticEntityType.GENE)) {
-            molecularDataListA = molecularDataService.fetchMolecularData(molecularProfileIdA, sampleIds, null, "SUMMARY");
+            molecularDataListA = molecularDataService.fetchMolecularData(molecularProfileIdA, sampleIds, null,
+                    "SUMMARY");
         } else if (geneticEntityType.equals(GeneticEntityType.GENESET)) {
             molecularDataListA = genesetDataService.fetchGenesetData(molecularProfileIdB, sampleIds, null);
         }
         MolecularProfile molecularProfileB = molecularProfileService.getMolecularProfile(molecularProfileIdB);
-        Boolean isMolecularProfileBOfGenesetType = molecularProfileB.getMolecularAlterationType().equals(MolecularProfile.MolecularAlterationType.GENESET_SCORE);
+        Boolean isMolecularProfileBOfGenesetType = molecularProfileB.getMolecularAlterationType()
+                .equals(MolecularProfile.MolecularAlterationType.GENESET_SCORE);
         if (isMolecularProfileBOfGenesetType) {
-            molecularDataListB = genesetDataService.fetchGenesetData(molecularProfileIdB, sampleIds, null).stream().collect(Collectors.toList());;
+            molecularDataListB = genesetDataService.fetchGenesetData(molecularProfileIdB, sampleIds, null).stream()
+                    .collect(Collectors.toList());
+            ;
         } else {
-            molecularDataListB = molecularDataService.fetchMolecularData(molecularProfileIdB, sampleIds, null, "SUMMARY").stream().collect(Collectors.toList());
+            molecularDataListB = molecularDataService
```

Comment:
so in this case (multiple profiles and NOT genesets, we lose our optimization, right, because we have get the molecularData (which were the original target of optimization).

Reply:
yes